### PR TITLE
Add support for Kafka's ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG (#1090)

### PR DESCRIPTION
**Pull request by MarvinCai**
Motivation:
Support configure Kafka's ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG in PulsarKafkaProducer by utilizing pulsar's ClientBuilder.keepAliveInterval.

### CODE REVIEW

**Comment by MarvinCai** [comment]
rerun java8 tests
rerun integration tests
rerun cpp tests

File: pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java [hunk 1]
```diff
@@ -114,7 +114,11 @@ private PulsarKafkaProducer(Map<String, Object> conf, Properties properties, Ser
 
         String serviceUrl = producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG).get(0);
         try {
-            client = PulsarClientKafkaConfig.getClientBuilder(properties).serviceUrl(serviceUrl).build();
+            // Support Kafka's ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG in ms.
+            long keepAliveIntervalMs = Long.parseLong(properties.getProperty(ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG, "30000"));
+            // If passed in value is greater than Integer.MAX_VALUE in second will throw ArithmeticException.
+            int keepAliveInterval = Math.toIntExact(keepAliveIntervalMs / 1000);
```
Comment (sijie) [inline pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java]: can you catch `ArithmeticException` and log an error message (e.g. "Invalid CONNECTIONS_MAX_IDLE_MS_CONFIG value - " + keepAliveIntervalMs + ". Please use a value smaller than " + Integer.MAX_VALUE") and then throw an InvalidArgumentException with the error message?

File: pulsar-client-kafka-compat/pulsar-client-kafka/src/test/java/org/apache/kafka/clients/producer/PulsarKafkaProducerTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        }).when(mockClientBuilder).keepAliveInterval(anyInt(), any(TimeUnit.class));
+
+        Properties properties = new Properties();
+        properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        properties.put(ProducerConfig.PARTITIONER_CLASS_CONFIG, DefaultPartitioner.class);
+        properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, Arrays.asList("pulsar://localhost:6650"));
+        properties.put(ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG, "1000000");
+
+        PulsarKafkaProducer<String, String> pulsarKafkaProducer = new PulsarKafkaProducer<>(properties, null, null);
```
Comment (sijie) [inline pulsar-client-kafka-compat/pulsar-client-kafka/src/test/java/org/apache/kafka/clients/producer/PulsarKafkaProducerTest.java]: Can you verify mockClientBuilder.keepAliveInterval is called once with the provided value?

**Comment by MarvinCai** [comment]
rerun java8 tests

**Comment by MarvinCai** [comment]
Addressed comments.

File: pulsar-client-kafka-compat/pulsar-client-kafka/src/test/java/org/apache/kafka/clients/producer/PulsarKafkaProducerTest.java (hunk 2 above)
Comment (MarvinCai) [reply pulsar-client-kafka-compat/pulsar-client-kafka/src/test/java/org/apache/kafka/clients/producer/PulsarKafkaProducerTest.java]: Verified mockClientBuilder.keepAliveInterval is called exactly once with expected value.

File: pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java (hunk 1 above)
Comment (MarvinCai) [reply pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java]: Done, catch ArithmeticException, log error message and throw IllegalArgumentException with same error message.

File: pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java [hunk 3]
```diff
@@ -34,6 +34,7 @@
 import java.util.concurrent.TimeoutException;
 import java.util.stream.Collectors;
 
+import com.sun.javaws.exceptions.InvalidArgumentException;
```
Comment (sijie) [inline pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java]: it seems that this is a wrong import?

Comment (MarvinCai) [reply pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java]: Removed unused import.
Also added document that merlimat mentioned.

**Comment by MarvinCai** [comment]
rerun java8 tests
rerun integration tests
rerun cpp tests

**Comment by MarvinCai** [comment]
rerun java8 tests
rerun integration tests
rerun cpp tests

**Comment by MarvinCai** [comment]
rerun java8 tests

**Comment by MarvinCai** [comment]
rerun java8 tests

**Comment by MarvinCai** [comment]
> This should be also updated in the consumer configuration and also we need to update the table with the supported configs in `site2/docs/adaptors-kafka.md`
Updated the description for this config in the table.

**Comment by MarvinCai** [comment]
rerun java8 tests

**Comment by MarvinCai** [comment]
rerun java8 tests

**Comment by MarvinCai** [comment]
rerun java8 tests

**Comment by MarvinCai** [comment]
rerun java8 tests

**Comment by sijie** [comment]
@merlimat can you review this PR again?

**Comment by MarvinCai** [comment]
rerun java8 tests
rerun integration tests

**Comment by MarvinCai** [comment]
rerun integration tests

**Comment by MarvinCai** [comment]
rerun integration tests

**Comment by sijie** [comment]
@MarvinCai thank you for your contribution!

**Comment by MarvinCai** [comment]
> @MarvinCai thank you for your contribution!
Thank you @sijie and @merlimat for reviewing the code and help me with this pr.
Will carry on and try contribute more  :smiley:

**Pull request merged**